**Coursebook: Python for Data Wrangling**
- Part 2 of Python Fundamental Course
- Course Length: 12 hours
- Last Updated: July 2019

___

- Developed by [Algoritma](https://algorit.ma)'s product division and instructors team

# Background

The coursebook is part of the **Python Fundamental Course** prepared by [Algoritma](https://algorit.ma). The coursebook is intended for a restricted audience only, i.e. the individuals and organizations having received this coursebook directly from the training organization. It may not be reproduced, distributed, translated or adapted in any form outside these individuals and organizations without permission.

Algoritma is a data science education center based in Jakarta. We organize workshops and training programs to help working professionals and students gain mastery in various data science sub-fields: data visualization, machine learning, data modeling, statistical inference etc.

## Training Objectives

On the second section of this **Python Fundamental Course**, we'll start exercising data wrangling practices using `pandas` libray. In this coursebook we will cover:

- Introduction to `pandas` library
- Exploratory Analysis Tools
- Group By Aggregation
- Multi-index Data Frame
- Data Tidying
- Connecting to SQL Source

By the end of this course, you'll be working on a **learn-by-building** module to create a data exploratory analysis project to apply what you have learned on provided dataset and attempt to answer all the given questions. This final part is considered as a Graded Assignment so make sure you do well.

# Introduction to pandas Library

We will start off by learning about a powerful Python data analysis library by the name of `pandas`. Its official documentation introduces itself as the "fundamental high-level building block for doing practical, real world data analysis in Python", and strive to do so by implementing many of the key data manipulation functionalities in R. This makes `pandas` a core member of many Python-based scientific computing environments.

From its [official documentation](https://pandas.pydata.org):

> Python has long been great for data munging and preparation, but less so for data analysis and modeling. pandas helps fill this gap, enabling you to carry out your entire data analysis workflow in Python without having to switch to a more domain specific language like R.

To use `pandas`, we will use Python's `import` function. Once imported, all `pandas` function can be accessed using the *pandas.function_name* notation.

In [1]:
import pandas as pd
print(pd.__version__)

0.24.2


## Introduction to DataFrame

Let's start with reading our first dataset, `amazon-electronic.csv`. The dataset is a sample sales data from electronic section of Amazon ecommerce.

In [2]:
elec = pd.read_csv("data_input/amazon-electronic.csv", index_col=0)
elec.head()

product_id        date              categories      brand  \
id                                                              
1        44342  2017-03-02          Camera & Photo  Panasonic   
2        46876  2017-03-02          Camera & Photo       Sony   
3        12136  2017-03-02          Camera & Photo       Sony   
6        79238  2017-03-03  Accessories & Supplies   Insignia   
7        46643  2017-03-03          Camera & Photo       Sony   

                                                 name     merchant  quantity  \
id                                                                             
1                       Lumix G 25mm f/1.7 ASPH. Lens  Bestbuy.com         1   
2         Cyber-shot DSC-WX220 Digital Camera (Black)  Bestbuy.com         1   
3               Sony - BC-TRX Battery Charger - Black  Bestbuy.com         1   
6   Insignia - Fixed TV Wall Mount For Most 40-70"...  Bestbuy.com         1   
7               Cyber-shot DSC-RX100 V Digital Camera  Bestbuy.com         1   

    unit_price  
id              
1       201.99  
2       159.99  
3        26.99  
6        56.99  
7       849.99

In the code above, we used `.read_csv()` to read a csv file from a specified path. Notice that we set `index_col=0` so the first column in the csv is used as the index. By default, this function treats the first row as the header row. We can add `header=None` to the function call telling `pandas` to read in a CSV without headers.

You may find it curious that we use `0` to reference the first element of an axis; This is because Python uses 0-based indexing, a behavior that is different from other languages such as R and Matlab.

**Discussion:**

Let's go further into understanding the `index_col` parameter. From the documentation:

> `index_col` : int or sequence or `False`  
Column to use as the row labels of the DataFrame.

1. How would you change the `read.csv()` code such that the DataFrame uses `product_id` as the row label (index)? 
2. `pandas.DataFrame.head()` accepts an additional parameter, `n`, and returns the first `n` rows of the DataFrame; Set `n=8` to see the first 8 rows of your `elect` DataFrame
3. The opposite of `.head()` is `.tail()`. It returns the last `n` row of your DataFrame. Create a new cell below and print the last 4 rows of our DataFrame 

*Reminder: Python uses 0-based indexing, and `product_id` is the second column in the csv*

In [3]:
## Your code below

## -- Solution code

### Keywords

Earlier it was mentioned that `index_col` accepts an integer, sequence or a `False` value. A couple of things to note here. `False`, along with its opposite, `True` are among a reserved list of vocabulary referred to as **Python Keywords**. We cannot use keyword as variable name, function name or assign values to them, essentially treating them as an identifier. 

Interestingly, all python keywords except **True**, **False** and **None** are in lowercase and they must be written as it is. As of Python 3.7 (latest version of Python as of this writing), there are 33 keywords:

`True`, `False`, `None`, `and`, `as`, `assert`, `break`, `class`, `continue`, `def`, `del`, `elif`, `else`
`except`, `finally`, `for`, `from`, `global`, `if`, `import`, `in`, `is`, `lambda`, `nonlocal`, `not`, `or`, `pass`
`raise`, `return`, `try`, `while`, `with`, `yield`

Try and insert a new cell below and assign the value of `2` to `False`. You should expect to see Python raising a `SyntaxError`:

> SyntaxError: can't assign to keyword

Which of the following 4 lines of code will evaluate without raising an error?

- [ ] `pd.read_csv("data_input/amazon-electronic.csv", index_col=false)`
- [ ] `Import pandas as pd`
- [ ] `print(100-2)`
- [ ] `None = 2`

You'll be tempted to go through all the keywords above and try to wrap your head around each one of them. If this proves to be a tad overwhelming, my recommendation is to move along the rest of the section; Most of us do not know the inner workings of every components of our car engine, but that shouldn't stop you from being an effective driver. 


As stated in the beginning of this course book, we're choosing a top-down approach and concepts will be presented on a "need-to-know" basis. We'll no doubt come across many of the keywords again (since collectively they form the backbone of the language) but for now, there is no need to stress about them if that only serve to discourage you from learning to code.

---

**What you need to know**:
- Python, as with `R`, `Swift`, `C`, and many other languages, are case-sensitive. `Sales` and `sales` refer to different objects.  
- You cannot use any Python keywords as identifers. 
- When naming your variables, start with a letter and use underscore (`_`) to join multiple words.
    - Wrong: `2019`, `2019sales`, `sales-2019`, `sales.2019`
    - Correct: `sales_2019`, `profit_after_tax`

`pandas` allow data analysts to create Series objects and DataFrame objects. Series is used to represent a one-dimensional array whereas DataFrame emulates the functionality of "Data Frames" in R and is useful for tabular data. 

In practice, a large proportion of our data is tabular: when we import data from a relational database (MySQL, Postgre) or from a spreadsheet software (Google Sheets, Microsoft Excel) we can represent these data as a DataFrame object.

## Exploratory Data Analysis Tools

In simple words, exploratory data analysis (EDA) refers to the process of performing initial investigations on data, often with the objective of becoming familiar with certain characteristics of the data. This is usually done with the aid of summary statistics and simple graphical techniques that purposefully uncover the structure of our data.

We'll start off by using some of the most convenient EDA tools conveniently built into `pandas`. Particularly, this is a summary of what we'll cover in common EDA workflows:

- `.head()` and `.tail()`
- `.describe()`
- `.shape` and `.size`
- `.axes`
- `.dtypes`

In [5]:
elec.describe()

product_id     quantity   unit_price
count   2519.000000  2519.000000  2519.000000
mean   38543.870584     1.154426   568.362060
std    22545.966776     0.468605   620.874159
min      366.000000     1.000000     1.000000
25%    18956.000000     1.000000   149.990000
50%    38926.000000     1.000000   479.750000
75%    56249.000000     1.000000   548.000000
max    81681.000000     6.000000  4199.990000

The `describe()` method will generate descriptive statistics of our data, and by default include all numeric columns in our DataFrame. The code above calls `.describe()` on `elect`, from which there are two numeric columns. This method is an "instruction" to perform something (functions) associated with the object. We've seen earlier how to use `.head()` and `.tail()` on our DataFrame: these are also method calls!

We can add an `include` parameter in the `.describe()` method call, which takes a list-like of dtypes to be included or `all` for all columns of the dataframe.

Add a new cell below, calling `describe()` but only on columns of `object` and `datetime` types (`['object', 'datetime']`).

In [6]:
## Your code below

## -- Solution code


Very often, we also want to know the shape of our data - i.e. how many rows and columns are there in our DataFrame? 

Our DataFrame has attributes that we can use to answer those questions. An attribute is a value stored within an object that describe an aspect of the object's characteristic. In the following call, we are asking for the `.shape` and the `.size` attribute of our `elect` DataFrame.

_Tip:_
Unlike `describe()`, which is a method call; `shape` and `size` are **attributes** of our DataFrame - that means no function is evaluated; Only a value stored in the object's instance is looked up and returned.

In [7]:
print(elec.shape)
print(elec.size)

(2519, 8)
20152


`size` returns the number of elements in the `elec` DataFrame. Because we have 2,519 rows and 8 columns, the total number of elements would be a total of 28416. 

Use `.shape` on the `member` DataFrame. From the resulting output, could you tell what would be the result of calling `member.size`?

In [8]:
## Your code below

## -- Solution code

One other attribute that is often useful is `.axes`, which return a list representing the axes of our DataFrame. Most likely, this would be a list of length 2, one for the row axis and one for the column axis, in that particular order.

Because it is ordered that way, calling `.axes[0]` would return the first item of that list, which would be the row axis (or row names if present) and calling `.axes[1]` would return the column axis, which would be equivalent to calling `elec.columns`:

In [9]:
elec.axes[1]

Index(['product_id', 'date', 'categories', 'brand', 'name', 'merchant',
       'quantity', 'unit_price'],
      dtype='object')

The last one is `.dtypes`, in which will show you the data types from the values in each column. To get better explanation on `.dtypes` and the data types in `pandas`, we'll go through the next chapter of this course, **Data Types.**

## Data Types

When we call `pd.read_csv()` earlier, `pandas` will try to infer data types from the values in each column. Sometimes, it get it right but more often that not, a data analyst's intervention is required. In the following sub-section, we'll learn about various techniques an analyst have at his/her disposal when it comes to the treatment of pandas data types.

In [10]:
# print(elec.dtypes)

elec.dtypes

product_id      int64
date           object
categories     object
brand          object
name           object
merchant       object
quantity        int64
unit_price    float64
dtype: object

`dtypes` simply stands for "data types". Because `elec` is a `pandas` object, accessing the `dtypes` attribute will return a series with the data type of each column. 

----

Look at the following code - what is the expected output from the following code? Why?
```
x = [2019, 4, 'data science']
x.dtypes
```

Hint: Try `type(x)` and verify the type for object `x`.

In [11]:
## Your code below

## -- Solution code

Let's take a look at some examples of `DataFrame.dtypes`:

In [12]:
member = pd.DataFrame({
    'name': ['Anita', 'Brian'],
    'birth': [pd.Timestamp('19931108'), pd.Timestamp('19800612')],
    'gender': pd.Categorical(['F','M']),
    'vip': [True, False],
    'ordercount': [11, 7],
    'avgbuy': [250554.32,500004.23 ]
})
member

name      birth gender    vip  ordercount     avgbuy
0  Anita 1993-11-08      F   True          11  250554.32
1  Brian 1980-06-12      M  False           7  500004.23

In [13]:
member.dtypes

name                  object
birth         datetime64[ns]
gender              category
vip                     bool
ordercount             int64
avgbuy               float64
dtype: object

Let's go through the columns and their data types from the above `DataFrame`:

- `name` [`object`]: store text values
- `birth` [`int`]: date and time values
- `gender`[`category`]: store categorical values
- `vip` [`bool`]: True/False values
- `ordercount` [`datetime`]: integer values
- `avgbuy` [`float`]: floating point values

Among these columns, only `ordercount` and `avgbuy` are columns with numeric values. This is a simple, but important, observation to make as we make our way into the Exploratory Data Analysis phase. But before we do, let's do one more exercise. Take a closer look at the Data Frame we just created again.

Both `name` and `gender` is not numerical, hence, `name` was stored as `object` while the `gender` store as `category`.

### Categorical and Numerical Variables

When working with categories, it is recommended both from a business point of a view and a technical one to use `pandas` categorical data type. From a business perspective, this adds clarity to the analyst's mind about the type of data he/she is working with. This informs and guides the analysis, on questions such as which statistical methods or plot types to use.

From a technical viewpoint, the memory savings -- and in turn, computation speed as well as computational resources -- can be quite significant. Specifically, the docs remarked:

> The memory usage of a `Categorical` is proportional to the number of categories plus the length of the data. In contrast, an `object` dtype is a constant times the length of the data

Now, let's get back to our `elec` data. Can you spot which of our column holds values that should be encoded in the `category` data type? Once you've spotted it, use the `astype('category')` method to perform the conversion. Remember to re-assign this new column so the original column (`object`) type is overwritten with the new `category` type column.

In [14]:
elec.dtypes

product_id      int64
date           object
categories     object
brand          object
name           object
merchant       object
quantity        int64
unit_price    float64
dtype: object

In [15]:
## Your code below


**Solution code**:

```py
# convert single variable
elec['categories'] = elec['categories'].astype('category')

# convert multiple variables
elec[['categories','brand','merchant']] = elec[['categories','brand','merchant']].astype('category')

```

### Datetime Variables

Given the program's special emphasis on business-driven analytics, one data type of particular interest to us is the `datetime`. In the first part of this coursebook, we've seen an example of `datetime` in the section introducing data types (`member.birth`).

Now, let's take a look at the data types of our `DataFrame` again:

In [16]:
elec.dtypes

product_id      int64
date           object
categories     object
brand          object
name           object
merchant       object
quantity        int64
unit_price    float64
dtype: object

Notice that all columns are in the right data types, except for `date`. The correct data type for this column would have to be a `datetime`. 

If our datetime data was stored in ISO format (YYYY-MM-DD) we can apply the same `astype('datetime64')` method. Or, another way to convert a column to a datetime is to use:

    `x = pd.to_datetime(x)`
    

In [17]:
# elec['date'].astype('datetime64')

elec['date'] = pd.to_datetime(elec['date'])

In fact, `pandas` has a number of machineries to work with `datetime` objects. These are convenient for when we need to extract the `month`, or `year`, or `weekday_name` from `datetime`. Some common applications in business analysis include:

- `elec['date'].dt.month`
- `elec['date'].dt.year`
- `elec['date'].dt.day`
- `elec['date'].dt.dayofweek`
- `elec['date'].dt.hour`
- `elec['date'].dt.weekday_name`

There are also other functions that can be helpful in certain situations. Supposed we want to transform the existing `datetime` column into values of periods we can use the `.to_period` method:

- `elec['date'].dt.to_period('D')`
- `elec['date'].dt.to_period('W')`
- `elec['date'].dt.to_period('M')`
- `elec['date'].dt.to_period('Q')`

For example, if we want to count how many sales records do we obtained within each quarter:

In [18]:
elec['date'].dt.to_period('Q').value_counts()

2017Q3    891
2017Q4    520
2018Q2    421
2017Q2    269
2017Q1    207
2018Q1    186
2018Q3     25
Freq: Q-DEC, Name: date, dtype: int64

## Indexing and Subsetting with Pandas

Using indexing operators to select, summarize or transform only a subset of data is a critical part of any data analysis workflow. Consider the following use-cases:

- Compare the sales in Year 2018 vs Year 2019  
- Identify missed opportunities in a specific market segment
- Best quarter of the year to execute cross-selling promos / discounts
- Study profitability of goods in the higher price range (e.g. IDR45000000+) and how competitors positioning affect sales in that price range

Notice that in all of these use-cases, data analysts will want to use some combination of indexing and then perform the necessary computations on that specific slice or slices of data. Unsurprisingly, `pandas` come with a number of methods to help you accomplish this task.

In the following section, we'll take a closer look at some of the most common slicing and subsetting operations in `pandas`:
- `head()` and `tail()`  
- `select_dtypes()`  
- Using `.drop()` 
- The `[]` operator
- `.loc`  
- `.iloc`
- Conditional subsetting

Say we're only really interested in the numeric columns of our data, we can use `select_dtypes` to selectively include or exclude only particular data types.

In the following example, I use `select_dtypes` to _include_ only textual columns (`objects`) and then proceed to pass the output of this function call into `.head()`. Notice that when we chain two methods this way, the output of the first function call will be "passed" into the second function call: 

In [19]:
elec.select_dtypes(include = 'object').head()

categories      brand  \
id                                      
1           Camera & Photo  Panasonic   
2           Camera & Photo       Sony   
3           Camera & Photo       Sony   
6   Accessories & Supplies   Insignia   
7           Camera & Photo       Sony   

                                                 name     merchant  
id                                                                  
1                       Lumix G 25mm f/1.7 ASPH. Lens  Bestbuy.com  
2         Cyber-shot DSC-WX220 Digital Camera (Black)  Bestbuy.com  
3               Sony - BC-TRX Battery Charger - Black  Bestbuy.com  
6   Insignia - Fixed TV Wall Mount For Most 40-70"...  Bestbuy.com  
7               Cyber-shot DSC-RX100 V Digital Camera  Bestbuy.com

Change the following code from `include` to `exclude` and observe the difference in the output from our `.describe()` call:

In [20]:
elec.select_dtypes(exclude='object').describe()

product_id     quantity   unit_price
count   2519.000000  2519.000000  2519.000000
mean   38543.870584     1.154426   568.362060
std    22545.966776     0.468605   620.874159
min      366.000000     1.000000     1.000000
25%    18956.000000     1.000000   149.990000
50%    38926.000000     1.000000   479.750000
75%    56249.000000     1.000000   548.000000
max    81681.000000     6.000000  4199.990000

You can also use `include` or `exclude` with a list of data types instead of a singular value. To include all columns of data types integer and float, we can do either of these:
- `include='number'`  
- `include=['int', 'float']`

Try and do that now; Chain the `select_dtypes()` command with `.head()` to limit the output to only the first 5 rows:

Apart from using `select_dtypes` to exclude columns, we can also use `.drop()` to remove rows or columns by label names and the corresponding axis. By default, the `axis` is assumed to be 0, i.e. referring to the row. Hence the following code will drop the **row** with label `1`: 

In [21]:
elec.drop(1).head()

product_id       date              categories     brand  \
id                                                            
2        46876 2017-03-02          Camera & Photo      Sony   
3        12136 2017-03-02          Camera & Photo      Sony   
6        79238 2017-03-03  Accessories & Supplies  Insignia   
7        46643 2017-03-03          Camera & Photo      Sony   
8         9472 2017-03-03          Camera & Photo    Yamaha   

                                                 name     merchant  quantity  \
id                                                                             
2         Cyber-shot DSC-WX220 Digital Camera (Black)  Bestbuy.com         1   
3               Sony - BC-TRX Battery Charger - Black  Bestbuy.com         1   
6   Insignia - Fixed TV Wall Mount For Most 40-70"...  Bestbuy.com         1   
7               Cyber-shot DSC-RX100 V Digital Camera  Bestbuy.com         1   
8       R-S202 Stereo Receiver with Bluetooth (Black)  Bestbuy.com         1   

    unit_price  
id              
2       159.99  
3        26.99  
6        56.99  
7       849.99  
8       125.99

We can drop multiple rows or columns by passing in a list. In the following code, we override the default `axis` value by passing `axis=1`; As a result `pandas` will drop the specified columns, while preserving all rows:

In [22]:
elec.drop(['product_id', 'brand', 'name'],axis = 1).head()

date              categories     merchant  quantity  unit_price
id                                                                      
1  2017-03-02          Camera & Photo  Bestbuy.com         1      201.99
2  2017-03-02          Camera & Photo  Bestbuy.com         1      159.99
3  2017-03-02          Camera & Photo  Bestbuy.com         1       26.99
6  2017-03-03  Accessories & Supplies  Bestbuy.com         1       56.99
7  2017-03-03          Camera & Photo  Bestbuy.com         1      849.99

Rather commonly, you may want to perform subsetting by slicing out a set of rows. This can be done using the `elec[start:end]` syntax, where `start` is inclusive.

The code follows slices out the first to fourth row, or equivalently, row with the index 0, 1, 2, and 3. 

In [23]:
elec[0:4]

product_id       date              categories      brand  \
id                                                             
1        44342 2017-03-02          Camera & Photo  Panasonic   
2        46876 2017-03-02          Camera & Photo       Sony   
3        12136 2017-03-02          Camera & Photo       Sony   
6        79238 2017-03-03  Accessories & Supplies   Insignia   

                                                 name     merchant  quantity  \
id                                                                             
1                       Lumix G 25mm f/1.7 ASPH. Lens  Bestbuy.com         1   
2         Cyber-shot DSC-WX220 Digital Camera (Black)  Bestbuy.com         1   
3               Sony - BC-TRX Battery Charger - Black  Bestbuy.com         1   
6   Insignia - Fixed TV Wall Mount For Most 40-70"...  Bestbuy.com         1   

    unit_price  
id              
1       201.99  
2       159.99  
3        26.99  
6        56.99

Recalling that the `end` is not inclusive and Python's 0-based indexing behavior, if we have wanted to subset the **8th to 12th** row of our data, how would we have done it instead? Pick the right answer and try it in a new code cell below.

- [ ] `elec[7:12]`
- [ ] `elec[8:12]`
- [ ] `elec[7:13]`
- [ ] `elec[8:13]`

Using `.loc` and `.iloc`, we can perform slicing on both the row and column indices, offering us even greater flexibility and control over our subsetting operations.

`.iloc` requires us to pass an `integer` to either the row or/and column. We can also use `:` to indicate no subsetting in a certain direction. The following code slices out the first 4 rows but take all columns (pay attention to the use of the `:` operator): 

In [24]:
elec.iloc[0:4, :]

product_id       date              categories      brand  \
id                                                             
1        44342 2017-03-02          Camera & Photo  Panasonic   
2        46876 2017-03-02          Camera & Photo       Sony   
3        12136 2017-03-02          Camera & Photo       Sony   
6        79238 2017-03-03  Accessories & Supplies   Insignia   

                                                 name     merchant  quantity  \
id                                                                             
1                       Lumix G 25mm f/1.7 ASPH. Lens  Bestbuy.com         1   
2         Cyber-shot DSC-WX220 Digital Camera (Black)  Bestbuy.com         1   
3               Sony - BC-TRX Battery Charger - Black  Bestbuy.com         1   
6   Insignia - Fixed TV Wall Mount For Most 40-70"...  Bestbuy.com         1   

    unit_price  
id              
1       201.99  
2       159.99  
3        26.99  
6        56.99

`.loc`, in contrast to `.iloc` does not subset based on _integer_ but rather subset based on `label`. We can still use `integer` but our integers will be treated or interpreted as _labels_.

Say, we're now wanted to use the `product_id` as the row labels:

In [25]:
product = elec.set_index('product_id')

product.head()

date              categories      brand  \
product_id                                                 
44342      2017-03-02          Camera & Photo  Panasonic   
46876      2017-03-02          Camera & Photo       Sony   
12136      2017-03-02          Camera & Photo       Sony   
79238      2017-03-03  Accessories & Supplies   Insignia   
46643      2017-03-03          Camera & Photo       Sony   

                                                         name     merchant  \
product_id                                                                   
44342                           Lumix G 25mm f/1.7 ASPH. Lens  Bestbuy.com   
46876             Cyber-shot DSC-WX220 Digital Camera (Black)  Bestbuy.com   
12136                   Sony - BC-TRX Battery Charger - Black  Bestbuy.com   
79238       Insignia - Fixed TV Wall Mount For Most 40-70"...  Bestbuy.com   
46643                   Cyber-shot DSC-RX100 V Digital Camera  Bestbuy.com   

            quantity  unit_price  
product_id                        
44342              1      201.99  
46876              1      159.99  
12136              1       26.99  
79238              1       56.99  
46643              1      849.99

To subset for the row of transactions corresponding to product id 62320, we can use label-based indexing (`.loc`) as such:

In [26]:
product.loc[70524, :]

date  categories   brand  \
product_id                                  
70524      2017-08-19  Home Audio  Yamaha   
70524      2017-08-22  Home Audio  Yamaha   
70524      2017-08-29  Home Audio  Yamaha   
70524      2017-11-18  Home Audio  Yamaha   
70524      2017-12-04  Home Audio  Yamaha   
70524      2017-12-06  Home Audio  Yamaha   
70524      2018-01-09  Home Audio  Yamaha   
70524      2018-05-04  Home Audio  Yamaha   
70524      2018-06-01  Home Audio  Yamaha   
70524      2018-07-25  Home Audio  Yamaha   

                                                         name  \
product_id                                                      
70524       NS-SP1800BL 5.1-Channel Home Theater System (B...   
70524       NS-SP1800BL 5.1-Channel Home Theater System (B...   
70524       NS-SP1800BL 5.1-Channel Home Theater System (B...   
70524       NS-SP1800BL 5.1-Channel Home Theater System (B...   
70524       NS-SP1800BL 5.1-Channel Home Theater System (B...   
70524       NS-SP1800BL 5.1-Channel Home Theater System (B...   
70524       NS-SP1800BL 5.1-Channel Home Theater System (B...   
70524       NS-SP1800BL 5.1-Channel Home Theater System (B...   
70524       NS-SP1800BL 5.1-Channel Home Theater System (B...   
70524       NS-SP1800BL 5.1-Channel Home Theater System (B...   

                    merchant  quantity  unit_price  
product_id                                          
70524       bhphotovideo.com         1      149.95  
70524       bhphotovideo.com         1      149.95  
70524       bhphotovideo.com         1      149.95  
70524       bhphotovideo.com         1      149.95  
70524       bhphotovideo.com         1      149.95  
70524       bhphotovideo.com         2      149.95  
70524       bhphotovideo.com         1      149.95  
70524       bhphotovideo.com         1      149.95  
70524       bhphotovideo.com         1      149.95  
70524       bhphotovideo.com         1      149.95

### pandas Conditional Subsetting

Along with `.iloc` and `.loc`, probably the most helpful type of subsetting would have to be conditional subsetting.

With conditional subsetting, we select data based on criteria we specified:
- `.categories == 'Home Audio'` to select all transactions where format is Home Audio  
- `.unit_price >= 200` to select all transactions with unit price being equal to or greater than 200 USD. 
- `.quantity != 1` to select all transactions where quantity of purchase **is not** 1   

We can also use the `&` and `|` operators to join conditions.

For example:

`customer[(customer.gender == 'Male') & (customer.age > 21)]` subset any rows where customer is Female with age more than 21.

In [27]:
elec[elec.categories == 'Home Audio'].head()

product_id       date  categories    brand  \
id                                               
29       71402 2017-03-03  Home Audio    Denon   
30       33811 2017-03-03  Home Audio  Kenwood   
32       30441 2017-03-03  Home Audio    Onkyo   
33       45721 2017-03-03  Home Audio    Onkyo   
34        6210 2017-03-03  Home Audio  Pioneer   

                                                 name     merchant  quantity  \
id                                                                             
29                  DP-300F Fully Automatic Turntable  Bestbuy.com         1   
30  Kenwood KMM-BT315U Digital Media Receiver with...  Bestbuy.com         1   
32    Onkyo TX-NR555 7.2-Channel Network A/V Receiver  Bestbuy.com         1   
33  Onkyo TXNR757 7.2 Channel Wireless A/V Receive...  Bestbuy.com         1   
34    Elite SC-LX501 7.2-Channel Network A/V Receiver  Bestbuy.com         1   

    unit_price  
id              
29      296.99  
30       89.99  
32      303.99  
33      503.99  
34      899.99

**Discussion:**

In the cell below, write code using conditional subsetting and answer the following questions:

1. Say on May 26 last year, we're holding a one-day only big sale promotion. From all of our electronic categories, how many purchases do we have?
2. On that day, how many transactions do we have from `Computers & Accessories`?
3. From all transactions in our dataset, how many transactions came from `Walmart.com`?

_Tip_: You may find the `.shape` attribute convenient in extracting the number of rows / columns from a dataframe

In [28]:
## Your code below


## -- Solution code

## Referencing and Copying

In [29]:
SEO_budget = [1350, 2000, 1750]

SEO_budget

[1350, 2000, 1750]

In the cell above, we created a Python list, and our variable `SEO_budget` referenced that object accordingly.

In the following cell, we created `Socmed_budget`, and set `Socmed_budget = SEO_budget`. What do you think it does?

In [30]:
Socmed_budget = SEO_budget
print(Socmed_budget)

[1350, 2000, 1750]


Now, let's try to update our `Socmed_budget` by changing our first value to 1500 USD:

In [31]:
Socmed_budget[0] = 1500
print(Socmed_budget)

[1500, 2000, 1750]


So far, so reasonable. However, if we were to now check the values of that Python list as referenced by `SEO_budget`, we see that __the list has been updated__.

In [32]:
# print(Socmed_budget)
print(SEO_budget)

[1500, 2000, 1750]


The explanation is that when we execute the line `Socmed_budget = SEO_budget`, a new Python list **is not being created**. We truly have only one object, and that line only creates a new variable named `Socmed_budget` that references that very same object.

The behavior with using the `=` operator in Python differs from other programming languages and can be confusing to seasoned developers new to Python. 

The appropriate method instead is `.copy()`, which creates an actual copy of the python list. Notice that in the following code, there are two distinct Python objects, and changing the values in one do not affect the other:

In [33]:
SEO_budget = [1350, 2000, 1750]
Socmed_budget = SEO_budget.copy()

Socmed_budget[0] = 1500

print(SEO_budget)
print(Socmed_budget)

[1350, 2000, 1750]
[1500, 2000, 1750]


**Discussion**:

The data you will read in is `electronic_views.csv`, a small sample from a dataset corresponding to product view counts from electronic section of a e-commerce. Notice that the dataset has some formatting inconsistencies by design: The `prices_after` column has comma delimiter and the currency (`USD`) whereas related columns use values that has omitted the separator.

In [34]:
views = pd.read_csv("data_input/electronic_views.csv")
views.head()

brand                                               name  \
0      Razer     Blade Pro 17.3 4K Ultra HD Touch-Screen Laptop   
1       Sony  Alpha a6500 Mirrorless Digital Camera (Body Only)   
2    Samsung  Refurbished Samsung Curved 65 4K (2160P) Smart...   
3         LG  75 Class - LED - UJ6470 Series - 2160p - Smart...   
4  Panasonic  Leica DG Vario-Elmar 100-400mm f/4-6.3 ASPH. P...   

                categories prices_before prices_after prices_disc  \
0  Computers & Accessories   USD4619.989  USD4,199.99         10%   
1           Camera & Photo     USD2295.4  USD1,996.00         15%   
2               TV & Video  USD3672.8895  USD3,497.99          5%   
3               TV & Video     USD2636.4  USD2,197.00         20%   
4           Camera & Photo   USD1919.988  USD1,599.99         20%   

   availability  total_views  
0          True           13  
1          True           20  
2         False           28  
3          True           15  
4         False           23

Pay attention to the `prices_before` and `prices_after` columns. To perform arithmetic computations on the numeric columns, we have to drop the 'USD' currency string and treat these columns as numbers. We'll use the built-in `.replace()` method for this.

How do we apply that replace function? We can call `.apply(our_function)` on our `DataFrame`. What's interesting is that `our_function` could be any of `python` built-in functions, functions from third-party modules, or it could also be a list of functions:

In [35]:
views['total_views'].apply([max, min])

max    43
min    13
Name: total_views, dtype: int64

Back to removing the currency string from `prices_before` and `prices_after` using `.apply()` and `.replace()`. We could create our own function , name it `removeUSD` for example and then apply it the following way:

`views['prices_before'].apply(removeUSD)`

Writing functions is a topic that is more suited for a later time, and students new to the trade of programming in Python will be gradually introduced to this aspect of Python programming.

However, given the task at hand, this seems like a reasonable time to introduce **Lambdas**.



In [36]:
views['prices_before'] = views['prices_before'].apply(lambda x: x.replace('USD', ''))
views['prices_after'] = views['prices_after'].apply(lambda x: x.replace('USD', ''))

views.head()

brand                                               name  \
0      Razer     Blade Pro 17.3 4K Ultra HD Touch-Screen Laptop   
1       Sony  Alpha a6500 Mirrorless Digital Camera (Body Only)   
2    Samsung  Refurbished Samsung Curved 65 4K (2160P) Smart...   
3         LG  75 Class - LED - UJ6470 Series - 2160p - Smart...   
4  Panasonic  Leica DG Vario-Elmar 100-400mm f/4-6.3 ASPH. P...   

                categories prices_before prices_after prices_disc  \
0  Computers & Accessories      4619.989     4,199.99         10%   
1           Camera & Photo        2295.4     1,996.00         15%   
2               TV & Video     3672.8895     3,497.99          5%   
3               TV & Video        2636.4     2,197.00         20%   
4           Camera & Photo      1919.988     1,599.99         20%   

   availability  total_views  
0          True           13  
1          True           20  
2         False           28  
3          True           15  
4         False           23

In [37]:
views = pd.read_csv("data_input/electronic_views.csv")

views[['prices_before','prices_after']] = views[['prices_before','prices_after']].replace('[^\d.]+', '',regex=True)
views.head(3)

brand                                               name  \
0    Razer     Blade Pro 17.3 4K Ultra HD Touch-Screen Laptop   
1     Sony  Alpha a6500 Mirrorless Digital Camera (Body Only)   
2  Samsung  Refurbished Samsung Curved 65 4K (2160P) Smart...   

                categories prices_before prices_after prices_disc  \
0  Computers & Accessories      4619.989      4199.99         10%   
1           Camera & Photo        2295.4      1996.00         15%   
2               TV & Video     3672.8895      3497.99          5%   

   availability  total_views  
0          True           13  
1          True           20  
2         False           28

In [38]:
views.dtypes

brand            object
name             object
categories       object
prices_before    object
prices_after     object
prices_disc      object
availability       bool
total_views       int64
dtype: object

**Task 1**

You may have noticed that `prices_before` and `prices_after` are not stored in the right data type. You may use `views.dtypes` to double-confirm.  

The following code uses the `pd.to_numeric` method to transform these columns to numeric data types. 

```py
numeric_cols = ['prices_before', 'prices_after']
views[numeric_cols] = views[numeric_cols].apply(pd.to_numeric)
```

Add the transformation code and then now try to count the difference amounts between the prices before and after the discount. Use `head` or `tail` to peek at the resulting data frame.

**Task 2**

The following code prints the product with the maximum amount of `total_views`:

```py
views.loc[views['total_views'].idxmax(), ]
views.loc[views['total_views'] == views['total_views'].max(), ]
```

Some of the products viewed by customers is not currently available (See `availability` column). Among the products that is not currently available, which product has the highest product views?

_Tip_: Recall what you've learned about conditional subsetting.

**Task 3**

Combine `.select_dtypes` and `.describe` to print the summary statistics of **all numeric columns** in `views`. 

_Tip_: This can be done by passing `['int','float']` or simply `'number'` as a value to the `include` parameter

**Task 4**

Say, we would like to print a dataframe including only any product with price more than USD 1,000. Which are the products that has price exceeding USD 1,000 in the sampled data frame?

----

# Data Preparation and Exploration

About 60 years ago, John Tukey defined data analysis as the "procedures for analyzing data, techniques for interpreting the results of such procedures ... and all the machinery of mathematical statistics which apply to analyzing data". His championing of EDA encouraged the development of statsitical computing packages, especially S at Bell Labs (which later inspired R).

He wrote a book titled _Exploratory Data Analysis_ arguing that too much emphasis in statistics was placed on hypothesis testing (confirmatory data analysis) while not enough was placed on the discovery of the unexpected. 

> Exploratory data analysis isolates patterns and features of the data and reveals these forcefully to the analyst.

This course aims to present a selection of EDA techniques -- some developed by John Tukey himself -- but with a special emphasis on its application to modern business analytics.

In the following chapters, we'll expand our EDA toolset with the following additions:  

- Tables
- Cross-Tables and Aggregates
- Using `aggfunc` for aggregate functions
- Pivot Tables


In this section, we're going to work with `clothing_review.csv`. It is a women’s clothing e-commerce dataset revolving around the reviews written by customers. This dataset includes 23,486 rows and 9 feature variables. Each row corresponds to a customer review, and includes the variables:

- `clothing_id`: refers to the specific piece being reviewed.
- `age_group`: categorical variable of the reviewers age group.
- `review_text`: string variable for the title of the review.
- `rating`: integer variable for the product score granted by the customer from 1 Worst, to 5 Best.
- `recommended_ind`: binary variable stating where the customer recommends the product where 1 is recommended, 0 is not recommended.
- `positive_feedback_count`: positive integer documenting the number of other customers who found this review positive.
- `division_name`: categorical name of the product high level division.
- `category`: categorical name of the product category.
- `review_date`: date of the review added.

We'll start by reading our dataset in. Since we already have some basic information of our data, notice that in the following code, we're also passing `parse_dates` argument into it:

In [39]:
review = pd.read_csv("data_input/clothing_reviews.csv", parse_dates = ['review_date'])
review.head()

clothing_id age_group                                        review_text  \
0          767  25 to 34  Absolutely wonderful - silky and sexy and comf...   
1         1080  25 to 34  Love this dress!  it's sooo pretty.  i happene...   
2         1077  55 to 64  I had such high hopes for this dress and reall...   
3         1049  35 to 54  I love, love, love this jumpsuit. it's fun, fl...   
4          847  35 to 54  This shirt is very flattering to all due to th...   

   rating  recommended_ind  positive_feedback_count   division_name  category  \
0       4                1                        0       Initmates  Intimate   
1       5                1                        4         General   Dresses   
2       3                0                        0         General   Dresses   
3       5                1                        0  General Petite   Bottoms   
4       5                1                        6         General      Tops   

          review_date  
0 2017-03-30 06:00:00  
1 2017-03-10 22:00:00  
2 2017-03-04 10:00:00  
3 2017-03-03 14:00:00  
4 2017-12-14 06:00:00

The `parse_dates` argument converts the `datetime` object, so we won't need further datetime conversion.

On the following code, we only need to convert the remaining columns to the correct data types:

In [40]:
review[['age_group', 'division_name', 'category']] = review[['age_group', 'division_name', 'category']].astype('category')

review.dtypes

clothing_id                         int64
age_group                        category
review_text                        object
rating                              int64
recommended_ind                     int64
positive_feedback_count             int64
division_name                    category
category                         category
review_date                datetime64[ns]
dtype: object

## Tables

One of the simplest EDA toolkit is the frequency table (contingency tables) and cross-tabulation tables. It is highly familiar, convenient, and practical for a wide array of statistical tasks. The simplest form of a table is to display counts of a `categorical` column.

In `pandas`, each column of a `DataFrame` is a `Series`. To get the counts of each unique levels in a categorical column, we can use `.value_counts()`. The resulting object is a `Series` and in descending order so that the most frequent element is on top. 

Suppose we want to know the amount of reviews obtained by each clothing `category`. Now, try and perform `.value_counts()` on the `category` column, adding either:

- `sort=False` as a parameter to prevent any sorting of elements, or
- `ascending=True` as a parameter to sort in ascending order instead

In [41]:
review.category.value_counts()

Tops        10468
Dresses      6319
Bottoms      3799
Intimate     1735
Jackets      1032
Trend         119
Name: category, dtype: int64

In [42]:
## Your code below


## -- Solution code

`crosstab` is a very versatile solution to producing frequency tables on a `DataFrame` object. Its utility really goes further than that but we'll start with a simple use-case.

Consider the following code: we use `pd.crosstab()` passing in the values to group by in the rows (`index`) and columns (`columns`) respectively. We're also setting the row (index) to be `category` and the function will by default compute a frequency table. 

In [43]:
pd.crosstab(index = review['category'],columns = "count")

col_0     count
category       
Bottoms    3799
Dresses    6319
Intimate   1735
Jackets    1032
Tops      10468
Trend       119

We can aslo add `normalize` argument on our crosstab to set the values to be normalized over each columns, and this will divide each values in place over the sum of all values.

In [44]:
pd.crosstab(index = review['category'], columns = "count", normalize = 'columns')

# equivalent to:
# cat = pd.crosstab(index = review['category'], columns = 'counts')
# cat/cat.sum()

col_0        count
category          
Bottoms   0.161852
Dresses   0.269214
Intimate  0.073918
Jackets   0.043967
Tops      0.445978
Trend     0.005070

We can also use the same `crosstab` method to compute a cross-tabulation of two factors. Say, for example, we want to know how our customer's age group corresponds to the reviews in each clothing category. 

So, in the following cell, we use `crosstab` with the `index` references the `category` column while the `columns` references the `age_group`. The `margins=True` on our method call adds an extra row and column of margins (subtotals):

In [45]:
pd.crosstab(index = review['category'], columns = review['age_group'], margins = True)

age_group  18 to 24  25 to 34  35 to 54  55 to 64   65+    All
category                                                      
Bottoms         130       815      2156       514   184   3799
Dresses         273      1524      3464       769   289   6319
Intimate         80       483       914       161    97   1735
Jackets          42       237       515       157    81   1032
Tops            360      2098      5747      1536   727  10468
Trend             6        15        75        15     8    119
All             891      5172     12871      3152  1386  23472

**Discussion:**

How many reviews obtained on each day of the week? Use `pd.crosstab(index=__, columns="count")` or `x.value_counts()`.

_Extra Tips_:
1. Use `x.dt.weekday_name`, assuming `x` is a datetime object to get the day of week. Assign this to a new column in your `review` Data Frame, name it `weekday`
2. Print the first 5 rows of your data to verify that your preprocessing steps are correct


In [46]:
## Your code below


## -- Solution code

If you've managed the above exercises, well done! Run the following cell anyway to make sure we're at the same starting point as we go into the next chapter.

In [47]:
review['weekday'] = review['review_date'].dt.weekday_name
review.head()
pd.crosstab(index = review['weekday'], columns = "count")

col_0      count
weekday         
Friday      3157
Monday      3015
Saturday    3528
Sunday      1569
Thursday    3468
Tuesday     5032
Wednesday   3717

### Aggregation Table

In the following section, we will introduce another parameter to perform aggregation on our table. The `aggfunc` parameter when present, required the `values` parameter to be specified as well. `values` is the values to aggregate according to the factors in our index and columns:

In [48]:
pd.crosstab(index=review['category'], 
            columns='Total Positive Feedbacks', 
            values=review['positive_feedback_count'],
            aggfunc='sum')

col_0     Total Positive Feedbacks
category                          
Bottoms                       8043
Dresses                      19510
Intimate                      3275
Jackets                       2916
Tops                         25407
Trend                          401

### Higher-dimensional Tables

If we need to inspect our data in higher resolution, we can create cross-tabulation using more than one factor. This allows us to yield insights on a more granular level yet have our output remain relatively compact and structured:

In [49]:
pd.crosstab(index = review['age_group'], 
            columns = [review['division_name'], review['category']], 
            values = review['rating'],
            aggfunc='mean')

division_name   General                                          \
category        Bottoms   Dresses   Jackets      Tops     Trend   
age_group                                                         
18 to 24       4.386364  4.311688  4.407407  4.394619  4.400000   
25 to 34       4.221402  4.089362  4.189542  4.107581  3.461538   
35 to 54       4.251366  4.167668  4.253918  4.101067  3.894737   
55 to 64       4.381381  4.218415  4.214286  4.292339  3.538462   
65+            4.295652  4.228070  4.270833  4.227926  4.000000   

division_name General Petite                                          \
category             Bottoms   Dresses  Intimate   Jackets      Tops   
age_group                                                              
18 to 24            4.404762  4.310924  4.250000  4.400000  4.226277   
25 to 34            4.234432  4.111301  4.108696  4.083333  4.106592   
35 to 54            4.346821  4.112551  4.226027  4.326531  4.207812   
55 to 64            4.320442  4.274834  4.619048  4.355932  4.305147   
65+                 4.507246  3.957627  4.250000  4.606061  4.408333   

division_name           Initmates  
category          Trend  Intimate  
age_group                          
18 to 24       3.000000  4.277778  
25 to 34       4.000000  4.265446  
35 to 54       3.777778  4.309896  
55 to 64       4.000000  4.278571  
65+                 NaN  4.200000

### Pivot Tables

If our data is already in a `DataFrame` format, using `pd.pivot_table` can sometimes be more convenient compared to a `pd.crosstab`. 

Fortunately, much of the parameters in a `pivot_table()` function is the same as `pd.crosstab()`. The noticable difference is the use of an additional `data` parameter, which allow us to specify the `DataFrame` that is used to construct the pivot table.

We create a `pivot_table` by passing in the following:
- `data`: our `DataFrame`
- `index`: the column to be used as rows
- `columns`: the column to be used as columns
- `values`: the values used to fill in the table
- `aggfunc`: the aggregation function

In [50]:
pd.pivot_table(
    data = review,
    index = 'age_group',
    columns = ['division_name', 'category'],
    values = 'rating',
    aggfunc = 'mean'
)

division_name   General                                          \
category        Bottoms   Dresses   Jackets      Tops     Trend   
age_group                                                         
18 to 24       4.386364  4.311688  4.407407  4.394619  4.400000   
25 to 34       4.221402  4.089362  4.189542  4.107581  3.461538   
35 to 54       4.251366  4.167668  4.253918  4.101067  3.894737   
55 to 64       4.381381  4.218415  4.214286  4.292339  3.538462   
65+            4.295652  4.228070  4.270833  4.227926  4.000000   

division_name General Petite                                          \
category             Bottoms   Dresses  Intimate   Jackets      Tops   
age_group                                                              
18 to 24            4.404762  4.310924  4.250000  4.400000  4.226277   
25 to 34            4.234432  4.111301  4.108696  4.083333  4.106592   
35 to 54            4.346821  4.112551  4.226027  4.326531  4.207812   
55 to 64            4.320442  4.274834  4.619048  4.355932  4.305147   
65+                 4.507246  3.957627  4.250000  4.606061  4.408333   

division_name           Initmates  
category          Trend  Intimate  
age_group                          
18 to 24       3.000000  4.277778  
25 to 34       4.000000  4.265446  
35 to 54       3.777778  4.309896  
55 to 64       4.000000  4.278571  
65+                 NaN  4.200000

A key difference between `crosstab` and `pivot_table` is that `crosstab` uses `len` (or `count`) as the default aggregation function while `pivot_table` using the mean. Copy the code from the cell below and make a change: use `sum` as the aggregation function instead: 

In [51]:
pd.pivot_table(
    data = review,
    index = 'category',
    columns= 'age_group',
    values = 'recommended_ind'
)

age_group  18 to 24  25 to 34  35 to 54  55 to 64       65+
category                                                   
Bottoms    0.861538  0.831902  0.848794  0.873541  0.896739
Dresses    0.864469  0.791339  0.807448  0.825748  0.806228
Intimate   0.850000  0.840580  0.853392  0.857143  0.855670
Jackets    0.904762  0.780591  0.856311  0.834395  0.839506
Tops       0.866667  0.787893  0.806160  0.850911  0.863824
Trend      1.000000  0.666667  0.720000  0.733333  0.875000

In [52]:
## Your code below


## -- Solution code

In [53]:
review.isna().sum()

clothing_id                  0
age_group                    0
review_text                845
rating                       0
recommended_ind              0
positive_feedback_count      0
division_name               14
category                    14
review_date                  0
weekday                      0
dtype: int64

## Missing Values and Duplicates

In the following cell, I'm going to  use `reindex` to "inject" some rows where values don't exist (clothing id 1100 through 1105) and also set `math.nan` on one of the values for `weekday`. Notice from the output that between row 3 to 8 there are at least a few rows with missing data.

In [54]:
import math
x=[i for i in range(1100, 1105)]
x.insert(2,820)
x

[1100, 1101, 820, 1102, 1103, 1104]

In [103]:
import math
x=[i for i in range(2400, 2404)]
x.insert(2,910)
x

review2 = pd.read_csv("data_input/clothing_reviews.csv").head(5).copy()
review2 = review2.reindex(x)
review2 = pd.concat([review2, review.head(9)], sort=True)
review2.loc[1049, "review_date"] = math.nan
review2.iloc[3:9, ]

age_group  category  clothing_id   division_name  \
2402       NaN       NaN          NaN             NaN   
2403       NaN       NaN          NaN             NaN   
0     25 to 34  Intimate        767.0       Initmates   
1     25 to 34   Dresses       1080.0         General   
2     55 to 64   Dresses       1077.0         General   
3     35 to 54   Bottoms       1049.0  General Petite   

      positive_feedback_count  rating  recommended_ind         review_date  \
2402                      NaN     NaN              NaN                 NaT   
2403                      NaN     NaN              NaN                 NaT   
0                         0.0     4.0              1.0 2017-03-30 06:00:00   
1                         4.0     5.0              1.0 2017-03-10 22:00:00   
2                         0.0     3.0              0.0 2017-03-04 10:00:00   
3                         0.0     5.0              1.0 2017-03-03 14:00:00   

                                            review_text   weekday  
2402                                                NaN       NaN  
2403                                                NaN       NaN  
0     Absolutely wonderful - silky and sexy and comf...  Thursday  
1     Love this dress!  it's sooo pretty.  i happene...    Friday  
2     I had such high hopes for this dress and reall...  Saturday  
3     I love, love, love this jumpsuit. it's fun, fl...    Friday

During the data exploration and preparation phase, it is likely we come across some problematic details in our data. This could be the value of _-1_ for the _age_ column, a value of _blank_ for the _customer segment_ column, or a value of _None_ for the _loan duration_ column. All of these are examples of "untidy" data, which is rather common depending on the data collection and recording process in a company.

In `pandas`, we use `NaN` (not a number) to denote missing data; The equivalent for datetime is `NaT` but both are essentially compatible with each other. From the docs:
> The choice of using `NaN` internally to denote missing data was largely for simplicity and performance reasons. We are hopeful that NumPy will soon be able to provide a native NA type solution (similar to R) performant enough to be used in pandas.

We can use `isna()` and `notna()` to detect missing values. An example code is as below:

In [56]:
review2['review_date'].isna()

2400     True
2401     True
910      True
2402     True
2403     True
0       False
1       False
2       False
3       False
4       False
5       False
6       False
7       False
8       False
1049     True
Name: review_date, dtype: bool

A common way of using the `.isna()` method is to combine it with the subsetting methods we've learned in previous lessons:

In [57]:
review2[review2['review_date'].isna()]

age_group category  clothing_id division_name  positive_feedback_count  \
2400       NaN      NaN          NaN           NaN                      NaN   
2401       NaN      NaN          NaN           NaN                      NaN   
910        NaN      NaN          NaN           NaN                      NaN   
2402       NaN      NaN          NaN           NaN                      NaN   
2403       NaN      NaN          NaN           NaN                      NaN   
1049       NaN      NaN          NaN           NaN                      NaN   

      rating  recommended_ind review_date review_text weekday  
2400     NaN              NaN         NaT         NaN     NaN  
2401     NaN              NaN         NaT         NaN     NaN  
910      NaN              NaN         NaT         NaN     NaN  
2402     NaN              NaN         NaT         NaN     NaN  
2403     NaN              NaN         NaT         NaN     NaN  
1049     NaN              NaN         NaT         NaN     NaN

Go ahead and use `notna()` to extract all the rows where `weekday` column is not missing:

In [58]:
## Your code below


## -- Solution code

Another common use-case in missing values treatment is to count the number of `NAs` across each column:

In [59]:
review2.isna().sum()

age_group                  6
category                   6
clothing_id                6
division_name              6
positive_feedback_count    6
rating                     6
recommended_ind            6
review_date                6
review_text                6
weekday                    6
dtype: int64

When we are certain that the rows with `NA`s can be safely dropped, we can use `dropna()`, optionally specifying a threshold. By default, this method drops the row if any NA value is present (`how='any'`), but it can be set to do this only when all values are NA in that row (`how='all'`).

```
    # drops row if all values are NA
    review2.dropna(how='all')
    
    # drops row if it doesn't have at least 5 non-NA values
    review2.dropna(thresh=5) 
```

In [60]:
review2.dropna(thresh=6).head()

age_group  category  clothing_id   division_name  positive_feedback_count  \
0  25 to 34  Intimate        767.0       Initmates                      0.0   
1  25 to 34   Dresses       1080.0         General                      4.0   
2  55 to 64   Dresses       1077.0         General                      0.0   
3  35 to 54   Bottoms       1049.0  General Petite                      0.0   
4  35 to 54      Tops        847.0         General                      6.0   

   rating  recommended_ind         review_date  \
0     4.0              1.0 2017-03-30 06:00:00   
1     5.0              1.0 2017-03-10 22:00:00   
2     3.0              0.0 2017-03-04 10:00:00   
3     5.0              1.0 2017-03-03 14:00:00   
4     5.0              1.0 2017-12-14 06:00:00   

                                         review_text   weekday  
0  Absolutely wonderful - silky and sexy and comf...  Thursday  
1  Love this dress!  it's sooo pretty.  i happene...    Friday  
2  I had such high hopes for this dress and reall...  Saturday  
3  I love, love, love this jumpsuit. it's fun, fl...    Friday  
4  This shirt is very flattering to all due to th...  Thursday

When we have data where duplicated observations are recorded, we can use `.drop_duplicates()` specifying whether the first occurence or the last should be kept:

In [61]:
print(review2.shape)
print(review2.drop_duplicates(keep="first").shape)

(15, 10)
(10, 10)


**Discussion:**
1. Duplicates may mean a different thing from a data point-of-view and a business analyst's point-of-view. You want to be extra careful about whether the duplicates is an intended characteristic of your data, or whether it poses a violation to the business logic. 

    - a. A medical center collects anonymized heart rate monitoring data from patients. It has duplicate observations collected across a span of 3 months
    - b. An insurance company uses machine learning to deliver dynamic pricing to its customers. Each row contains the customer's name, occupation / profession and historical health data. It has duplicate observations collected across a span of 3 months
    - c. On our original `review` data, check for duplicate observations. Would you have drop the duplicated rows?

---


In [62]:
review[review.duplicated(keep = False)].shape

(4, 10)


2. Once you've identified the missing values, there are 3 common ways to deal with it:

    - a. Use `dropna` with a reasonable threshold to remove any rows that contain too little values rendering it unhelpful to your analysis
    - b. Replace the missing values with a central value (mean or median)
    - c. Imputation through a predictive model
        - In a dataframe where `salary` is missing but the bank has data about the customer's occupation / profession, years of experience, years of education, seniority level, age, and industry, then a machine learning model such as regression or nearest neighbor can offer a viable alternative to the mean imputation approach
 
Going back to `review2`: what is a reasonable strategy? List them down or in pseudo-code.

In [63]:
## Your code below


## -- Solution code

### Missing Values Treatment

Some common methods when working with missing values are demonstrated in the following section. We make a copy of the NA-included DataFrame, and name it `review3`:

In [64]:
review3 = review2.copy()
review3.head(7)
review3.dtypes

age_group                          object
category                           object
clothing_id                       float64
division_name                      object
positive_feedback_count           float64
rating                            float64
recommended_ind                   float64
review_date                datetime64[ns]
review_text                        object
weekday                            object
dtype: object

In the following cell, the technique is demonstrably repetitive or even verbose. This is done to give us an idea of all the different options we can pick from. 

You may observe, for example that the two lines of code are functionally identical:
- `.fillna(0)`
- `.replace(np.nan, 0)`

In [102]:
review3[['age_group','review_text', 'division_name', 'category']] = review3[['age_group','review_text', 'division_name', 'category']].fillna('Missing')
review3[['recommended_ind','positive_feedback_count']] = review3[['recommended_ind','positive_feedback_count']].fillna(0)
review3.review_date = review3.review_date.fillna(method='bfill')
review3.rating = review3.rating.fillna(review3.rating.median())

review3.head(10)

age_group  category  clothing_id   division_name  \
2400   Missing   Missing          NaN         Missing   
2401   Missing   Missing          NaN         Missing   
910    Missing   Missing          NaN         Missing   
2402   Missing   Missing          NaN         Missing   
2403   Missing   Missing          NaN         Missing   
0     25 to 34  Intimate        767.0       Initmates   
1     25 to 34   Dresses       1080.0         General   
2     55 to 64   Dresses       1077.0         General   
3     35 to 54   Bottoms       1049.0  General Petite   
4     35 to 54      Tops        847.0         General   

      positive_feedback_count  rating  recommended_ind         review_date  \
2400                      0.0     5.0              0.0 2017-03-30 06:00:00   
2401                      0.0     5.0              0.0 2017-03-30 06:00:00   
910                       0.0     5.0              0.0 2017-03-30 06:00:00   
2402                      0.0     5.0              0.0 2017-03-30 06:00:00   
2403                      0.0     5.0              0.0 2017-03-30 06:00:00   
0                         0.0     4.0              1.0 2017-03-30 06:00:00   
1                         4.0     5.0              1.0 2017-03-10 22:00:00   
2                         0.0     3.0              0.0 2017-03-04 10:00:00   
3                         0.0     5.0              1.0 2017-03-03 14:00:00   
4                         6.0     5.0              1.0 2017-12-14 06:00:00   

                                            review_text   weekday  
2400                                            Missing       NaN  
2401                                            Missing       NaN  
910                                             Missing       NaN  
2402                                            Missing       NaN  
2403                                            Missing       NaN  
0     Absolutely wonderful - silky and sexy and comf...  Thursday  
1     Love this dress!  it's sooo pretty.  i happene...    Friday  
2     I had such high hopes for this dress and reall...  Saturday  
3     I love, love, love this jumpsuit. it's fun, fl...    Friday  
4     This shirt is very flattering to all due to th...  Thursday

**Discussion:**
1. Duplicates may mean a different thing from a data point-of-view and a business analyst's point-of-view. You want to be extra careful about whether the duplicates is an intended characteristic of your data, or whether it poses a violation to the business logic. 

    - a. A medical center collects anonymized heart rate monitoring data from patients. It has duplicate observations collected across a span of 3 months
    - b. An insurance company uses machine learning to deliver dynamic pricing to its customers. Each row contains the customer's name, occupation / profession and historical health data. It has duplicate observations collected across a span of 3 months
    - c. On our original `review` data, check for duplicate observations. Would you have drop the duplicated rows?


In [66]:
review[review.duplicated(keep = False)].shape

(4, 10)


2. Once you've identified the missing values, there are 3 common ways to deal with it:

    - a. Use `dropna` with a reasonable threshold to remove any rows that contain too little values rendering it unhelpful to your analysis
    - b. Replace the missing values with a central value (mean or median)
    - c. Imputation through a predictive model
        - In a dataframe where `salary` is missing but the bank has data about the customer's occupation / profession, years of experience, years of education, seniority level, age, and industry, then a machine learning model such as regression or nearest neighbor can offer a viable alternative to the mean imputation approach
 
Going back to `review2`: what is a reasonable strategy? List them down or in pseudo-code.

In [67]:
## Your code below


## -- Solution code

3. Read `amazon-electronic.csv` from the `data_input` folder. You may find it helpful to use `parse_dates=[__]` in the `read_csv()` call. This is the same data we worked on in the previous course.

Perform the necessary data preparation steps and use the exploratory data analysis techniques you've acquired to answer the questions below. 
   - a. Which brand is the favorite among TV & Video customers? 
   - b. If we convert our record to monthly period, when was the highest sales occur?
   - c. On that peak period, which merchant had the highest _sales volume_?
   - d. List top 5 best selling camera in 2017!    

In [68]:
## Your code below


## -- Solution code

----

# Data Wrangling and Reshaping

In the previous sections, we've got our hands on a few common techniques and learned how to explore data using `pandas` built-in methods. Specifically, we've  in the first and second part of this series how to use the following inspection, diagnostic and exploratory tools: 

**Data Inspection**
- `.head()` and `.tail()`
- `.describe()`
- `.shape` and `.size`
- `.axes`
- `.dtypes`
- Subsetting using `.loc`, `.iloc` and conditionals
---
**Diagnostic and Exploratory**
- Tables
- Cross-Tables and Aggregates
- Using `aggfunc` for aggregate functions
- Pivot Tables
- Working with DateTime
- Working with Categorical Data
- Duplicates and Missing Value Treatment

The first half of this section serves as an extension from the last. We'll pick up some new techniques to supplement our EDA toolset. Let us begin with reshaping techniques. 

## Pickling pandas DataFrame

Python pickle module is used for serializing and de-serializing a Python object structure. Any object in Python can be pickled so that it can be saved on disk. Pandas also has a very easy to use pickling functions. First we’ll make an example dataframe:

In [69]:
traffic = pd.read_csv('data_input/la-traffic.csv')
traffic = pd.pivot_table(traffic, index=['Date', 'Device Category']).unstack()

Then, we serialize the DataFrame object to a byte stream using `pandas.DataFrame.to_pickle`:

In [70]:
traffic.to_pickle('data_input/traffic.pkl')

In the cell above, we were saving `traffic` dataframe to a pickle file called `traffic.pkl` in our `data_input/` folder in our directory. 

Now, for the purposes of demonstration again, I’ll delete the original DataFrame, then load the pickled dataframe with `pandas.DataFrame.read_pickle`:

In [71]:
del traffic
traffic = pd.read_pickle('data_input/traffic.pkl')
traffic.head()

Bounce Rate                           Sessions              \
Device Category     desktop     mobile     tablet      desktop      mobile   
Date                                                                         
2019-06-01        41.455099  64.462710  20.012573   337.921687  582.246269   
2019-06-02        32.372644  46.659636  44.833929   320.876543  916.636364   
2019-06-03        28.904192  43.962397  85.697323   821.010471  891.817610   
2019-06-04        38.309282  57.369383  22.644138  1006.187097  837.184211   
2019-06-05        52.500556  59.074780  22.220044   871.426136  647.034091   

                            Total Visitors                          
Device Category      tablet        desktop      mobile      tablet  
Date                                                                
2019-06-01       115.878788     314.909639  477.708955  115.878788  
2019-06-02       132.492308     305.611111  741.898990  131.892308  
2019-06-03       270.419355     744.178010  695.792453  115.870968  
2019-06-04       154.600000     908.735484  684.184211   64.477778  
2019-06-05       141.804598     788.664773  519.465909  128.896552

## Data Reshaping

### `stack()` and `unstack()`

`stack()` stack the prescribed level(s) from columns to index and is particularly useful on DataFrames having a multi-level columns. It does so by "shifting" the columns to create new levels on its index. 

This is easier understood when we just see an example. Notice that `traffic` has a 2-level column (Device Category and each traffic parameters) and 1-level index (Date):

In [72]:
traffic.tail()

Bounce Rate                           Sessions              \
Device Category     desktop     mobile     tablet      desktop      mobile   
Date                                                                         
2019-06-26        32.394944  48.672402  44.497656  1078.000000  835.851064   
2019-06-27        50.778923  55.180274  47.058824   953.480769  790.500000   
2019-06-28        42.412381  53.702514  55.248658   728.851064  902.729730   
2019-06-29        35.244142  48.952851  25.000000   210.684211  549.027027   
2019-06-30        40.833044  46.124542  39.668914   228.375000  584.307692   

                            Total Visitors                          
Device Category      tablet        desktop      mobile      tablet  
Date                                                                
2019-06-26        77.814815     894.437500  642.574468   64.814815  
2019-06-27        93.647059     813.153846  635.142857   75.470588  
2019-06-28        72.875000     604.893617  725.324324   61.625000  
2019-06-29        78.000000     192.315789  467.108108   58.500000  
2019-06-30       122.142857     196.843750  517.038462  103.714286

When we stack the `traffic` DataFrame, we shrink the number of levels on its column by one: `traffic` now has 1-level column which is `Device Category`: 

In [73]:
traffic.stack().head(8)

Bounce Rate    Sessions  Total Visitors
Date       Device Category                                         
2019-06-01 desktop            41.455099  337.921687      314.909639
           mobile             64.462710  582.246269      477.708955
           tablet             20.012573  115.878788      115.878788
2019-06-02 desktop            32.372644  320.876543      305.611111
           mobile             46.659636  916.636364      741.898990
           tablet             44.833929  132.492308      131.892308
2019-06-03 desktop            28.904192  821.010471      744.178010
           mobile             43.962397  891.817610      695.792453

`unstack()` does the opposite: it "shifts" the levels from index axis onto column axis. Try and create a stack DataFrame, and then apply `unstack` on the new DataFrame to see it return to the original shape:

In [74]:
## Your code below


## -- Solution code


**Discussion:**

Which of the following statement is correct?

- [ ] `stack()` changes the DataFrame from wide to long
- [ ] `unstack()` changes the DataFrame from long to wide
- [ ] `unstack()` changes the DataFrame from wide to long

### Melt

Speaking of reshaping a DataFrame from wide format to long, another method that should be in your toolset is `melt()`. Consider the following DataFrame, which is created from `pandas` MultiIndex Slicers method, `.xs()`:

In [75]:
mobile = traffic.xs(['mobile'], level = ['Device Category'], axis = 1)
mobile.head()

Bounce Rate    Sessions  Total Visitors
Date                                               
2019-06-01    64.462710  582.246269      477.708955
2019-06-02    46.659636  916.636364      741.898990
2019-06-03    43.962397  891.817610      695.792453
2019-06-04    57.369383  837.184211      684.184211
2019-06-05    59.074780  647.034091      519.465909

In [76]:
mobile.shape

(30, 3)

The DataFrame above is wide: it has 35 rows and 3 columns. The `melt()` function gathers all the columns into one and store the value corresponding to each column such that the resulting DataFrame has 35 * 3 = 74 rows, along with the identifier and values columns:

In [77]:
mobile_melted = mobile.melt()
mobile_melted.head()

variable      value
0  Bounce Rate  64.462710
1  Bounce Rate  46.659636
2  Bounce Rate  43.962397
3  Bounce Rate  57.369383
4  Bounce Rate  59.074780

In [78]:
mobile_melted.shape

(90, 2)

We can optionally specify one or more columns to be identifier variables (`id_vars`), which treat all other columns as value variables (`value_vars`):

In [79]:
mobile.reset_index().melt(id_vars='Date').head()

Date     variable      value
0  2019-06-01  Bounce Rate  64.462710
1  2019-06-02  Bounce Rate  46.659636
2  2019-06-03  Bounce Rate  43.962397
3  2019-06-04  Bounce Rate  57.369383
4  2019-06-05  Bounce Rate  59.074780

## Group By

Reshaping data is an important component of any data wrangling toolkit as it allows the analyst to "massage" the data into the desired shape for further processing. 

Another equally important technique is the group by operation. Analysts having some experience with SQL or other data analysis toolsets (R's `tidyverse` for example) will find the group by operation a familiar strategy in many analysis-heavy workflow.

Consider the following DataFrame:

In [96]:
# extract datetime values from index
mobile.index = pd.to_datetime(mobile.index)
mobile['Weekday'] = mobile.index.weekday_name
mobile['Weeknumber'] = mobile.index.week

# create 'Visitors Growth' variable
mobile['Visitors Growth'] = mobile['Total Visitors'].diff()
mobile = mobile.round(2)
mobile.tail()

Bounce Rate  Sessions  Total Visitors    Weekday  Weeknumber  \
Date                                                                       
2019-06-26        48.67    835.85          642.57  Wednesday          26   
2019-06-27        55.18    790.50          635.14   Thursday          26   
2019-06-28        53.70    902.73          725.32     Friday          26   
2019-06-29        48.95    549.03          467.11   Saturday          26   
2019-06-30        46.12    584.31          517.04     Sunday          26   

            Visitors Growth  
Date                         
2019-06-26           -37.09  
2019-06-27            -7.43  
2019-06-28            90.18  
2019-06-29          -258.21  
2019-06-30            49.93

Pay special attention to how the `Visitors Growth` column was created. It's the difference between the `Total Visitors` value of the web traffic on a given day and the following day.

Supposed we want to compare the `Visitors Growth` between each `Weekday`; On average, does the visitors either increasing or decreasing on weekends?

In [81]:
mobile.groupby('Weekday').mean()

Bounce Rate  Sessions  Total Visitors  Weeknumber  Visitors Growth
Weekday                                                                      
Friday         40.7625  666.6525        543.8975        24.5           5.8025
Monday         55.5275  837.1525        659.2650        24.5          95.5575
Saturday       55.0600  530.5520        444.1480        24.0        -108.1375
Sunday         46.6780  643.3920        554.3720        24.0         110.2220
Thursday       43.8775  669.6950        538.0900        24.5         -97.4375
Tuesday        54.1175  827.4225        677.4175        24.5          18.1500
Wednesday      55.0175  811.3625        635.5300        24.5         -41.8875

We can pass in multiple columns to the `.groupby()` call like so:

In [82]:
mobile.groupby(by = ['Weekday','Weeknumber']).mean().head(8)

Bounce Rate  Sessions  Total Visitors  Visitors Growth
Weekday Weeknumber                                                        
Friday  23                18.30    573.93          481.85            19.15
        24                46.84    536.12          421.89            11.45
        25                44.21    653.83          546.53           -97.57
        26                53.70    902.73          725.32            90.18
Monday  23                43.96    891.82          695.79           -46.11
        24                57.75    682.20          546.37           -19.51
        25                59.75    894.61          695.76           283.68
        26                60.65    879.98          699.14           164.17

Before we going into some practice, let's look at some **group by** strategies when dealing with MultiIndex DataFrames. Consider the following DataFrame:

In [83]:
traffic.stack().head(9)

Bounce Rate    Sessions  Total Visitors
Date       Device Category                                         
2019-06-01 desktop            41.455099  337.921687      314.909639
           mobile             64.462710  582.246269      477.708955
           tablet             20.012573  115.878788      115.878788
2019-06-02 desktop            32.372644  320.876543      305.611111
           mobile             46.659636  916.636364      741.898990
           tablet             44.833929  132.492308      131.892308
2019-06-03 desktop            28.904192  821.010471      744.178010
           mobile             43.962397  891.817610      695.792453
           tablet             85.697323  270.419355      115.870968

On the Columns, there is 1 level. On the Index, there are 2 levels: `Date` and `Device Category` respectively. Supposed we want to groupby the second level of the axis (`Device Category`), we can pass in `axis=0` (referring to the row) and `level=1` (referring to the second level of index). 

The following groups it by the device category and aggregate the values from the remaining levels using the `sum`: 

In [84]:
traffic.stack().groupby(level = 1, axis = 0).sum()

Bounce Rate      Sessions  Total Visitors
Device Category                                           
desktop          1166.914829  20366.955036    17909.948695
mobile           1505.916541  21118.878445    17209.416402
tablet           1187.790763   3293.201097     2701.731334

**Discussion:**

1. During June, which date has the highest Visitor growth for each desktop, mobile, and tablet? What about the highest total Visitor growth from our web in general?  
<br>
2. Read `data_input/techcrunch.csv`, a dataset that stores fundraising rounds and amounts from startup companies of different categories around the US. Use `groupby` and perform a `median` aggregation. Which `category` raised the most amount in funding on average (use the `median`):

    - [ ] `mobile`
    - [ ] `cleantech`
    - [ ] `biotech`
    - [ ] `consulting`
<br>
<br>
3. Among all companies in San Francisco, which of the following are not among the top 5 most funded (highest `raisedAmt`) companies?
<br>
    - Hint: You may need to subset for companies where `city` is `San Francisco`, then group by `company` and `sum` on `raisedAmt` within each group. If it's helpful, you may consider chaining `sort_values()` and set `ascending=False` to get the DataFrame sorted by `raisedAmt` from largest to smallest

    - [ ] `OpenTable`
    - [ ] `Friendster`
    - [ ] `Facebook`
    - [ ] `Snapfish`
    
----

# Working with SQL Databases

There are a great number of python modules that provide functionalities to work with databases of all variants and flavors. For a MySQL database, we may form a connection using `pymysql` or one of many other alternatives:

```
import pymysql
conn = pymysql.connect(
    host=host,
    port=port,
    user=user,
    password=password,
    db=database)
```

We can then use `pd.read_sql_query()`, passing in the connection:
```
sales = pd.read_sql_query("SELECT * FROM sales", conn)
```

Under the hood, `pandas` uses SQLAlchemy so any database supported by that library will work. This isn't something you need to worry about at this stage of your learning journey, but for the sake for practice, let's also see how a connection URI for a SQLite database looks like:

In [104]:
import sqlite3
conn = sqlite3.connect("data_input/chinook.db")

albums = pd.read_sql_query("SELECT * FROM albums", conn)
albums.head()

AlbumId                                  Title  ArtistId
0        1  For Those About To Rock We Salute You         1
1        2                      Balls to the Wall         2
2        3                      Restless and Wild         2
3        4                      Let There Be Rock         1
4        5                               Big Ones         3

In the above command, we asked for all columns of a table to be returned to us through the `SELECT *` command. Well, columns of which table? That would be `tables`. Together they form an SQL query:

`SELECT * FROM albums`

The database we're working with have a few tables populated with sample data. The database has the following schema:
![](assets/chinookschema.png)

**Discussion:**

We'll create a `DataFrame`: this time select all columns from the `artists` table. Recall that when we use `pd.read_sql_query()` command we pass in the SQL query as a string, and add a connection as the second parameter. Save the output as a `DataFrame`.

Your DataFrame should be constructed like this:

`__ = pd.read_sql_query("SELECT __ FROM __ ", conn)`

Question:
1. How many rows are there in your DataFrame?

In [86]:
## Your code below


## -- Solution code

The `pd.read_sql_query` is most commonly used with that two parameters above, but on its [official documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sql_query.html) is a list of other parameters we can use as well. 

In the following cell, we use a similar SQL query with an additional `LIMIT` statement to limit the output to the first 5 records (rows). However, notice that we also set `index_col` so the specified column is recognized as the index:

In [105]:
pd.read_sql_query("SELECT * FROM artists LIMIT 5", 
                  conn, 
                  index_col='ArtistId')

Name
ArtistId                   
1                     AC/DC
2                    Accept
3                 Aerosmith
4         Alanis Morissette
5           Alice In Chains

## SQL Joins

`JOIN` statements are used to combine records from two tables. We can have as many `JOIN` operations as we want in a SQL query.

Below is a diagram of the different types of SQL `JOIN` operations:

![](assets/sqljoins.png)

Credit: Data & Object Factory, LLC

In most business scenarios though, a `LEFT JOIN` is almost always the type of `JOIN` you want - it is very direct (and therefore easy to reason about). Left join return all records in the left table regardless if any of those records have a match in the right table.

The `INNER JOIN` is also very intuitive and easily understood. This query return all of the records in the left table that has a matching record in the right table.

> As a personal side note, I've worked at companies where `RIGHT JOIN` is outright forbidden in favor of `LEFT JOIN`: directness and ease-of-understanding aside, all right joins can be replaced by the opposite left join. 
> 
> The `OUTER JOIN` (also referred to as `FULL JOIN`) is also quite uncommon in practice. Performance reason aside, an outer join return all of the records from both tables regardless if there is a match or not, resulting in a DataFrame that has potentially a lot of `NULL` values.

Consider the database schema illustraation again and pay attention to two tables and their respective columns:

1. `albums`: 
    - `AlbumId`, `Title`, `ArtistId`

2. `artists`: `
    - `ArtistId`, `Name`    

We want a `pandas` DataFrame containing the `AlbumId`, `Title` and `Name`. Notice that `Name` is from the `artists` table while the other columns are from the `albums` table. What is a reasonable strategy?

The most straightforward solution is the `LEFT JOIN`, let's see an example:

In [106]:
albums = pd.read_sql_query("SELECT AlbumId, Title, a.Name \
                           FROM albums \
                           LEFT JOIN artists as a \
                           ON a.ArtistId = albums.ArtistId", conn)
albums.head()

AlbumId                                  Title       Name
0        1  For Those About To Rock We Salute You      AC/DC
1        2                      Balls to the Wall     Accept
2        3                      Restless and Wild     Accept
3        4                      Let There Be Rock      AC/DC
4        5                               Big Ones  Aerosmith

Notice that in the code above, we place a backslash (`\`) character so we have line continuation and the newline will be ignored. This allows SQL to treat the entire query string as if they were essentially one line.

In [107]:
pd.read_sql_query("SELECT * FROM albums", conn).head()

AlbumId                                  Title  ArtistId
0        1  For Those About To Rock We Salute You         1
1        2                      Balls to the Wall         2
2        3                      Restless and Wild         2
3        4                      Let There Be Rock         1
4        5                               Big Ones         3

**Discussion:**

Consider the database schema illustration again and pay attention to two tables and their respective columns:

1. `albums`: `AlbumId`, `Title`, `ArtistId`

2. `tracks`: `TrackId`, `Name`, `AlbumId`, `GenreId`, ... `UnitPrice` 
    
3. `genres`: `GenreId`, `Name`

Create a `DataFrame` containing all columns from the `tracks` table; Additionally, it should also contain:
    - The `Title` column from the `albums` table
    - The `Name` column from the `artists` table
    - The `Name` column from the `genres` table 

> **Hint 1**: In your `SELECT` statement, you can use `SELECT tracks.* FROM TRACKS` to select all columns from the `TRACKS` table
> 
> **Hint 2**: When we write `SELECT tracks.Name as tracksName`, we are renaming the output column from `Name` to `tracksName` using a technique called column aliasing. You may optionally consider doing this for columns that share the same name across different tables 

Set the `TrackId`column to be the index. The resulting `DataFrame` should has 11 columns.

Give your `DataFrame` a name: name it `tracks`. Perform EDA on `tracks` to answer the following question:

1. Use `tail()` to inspect the last 5 rows of data. Which genre is present in the last 5 rows of our `tracks` DataFrame (Check all that apply)?
    - [ ] Latin
    - [ ] Classical
    - [ ] Soundtrack
    - [ ] Pop
  
  
2. Apply `pd.crosstab(..., columns='count')`, `.value_counts()`, or any other techniques you've learned to compute the frequency table of Genres in your DataFrame. Which is among the top 3 most represented genres in the `tracks` DataFrame?
    - [ ] Latin
    - [ ] Classical
    - [ ] Soundtrack
    - [ ] Pop
  
  
3. Use `groupby()` on Artist Name and compute the `mean()` on the `UnitPrice` of each tracks. You will realize that most artists price their tracks at 0.99 (`mean`) but there are several artists where the `mean()` is 1.99. Which of the Artist has a mean of 0.99 `UnitPrice`:
    - [ ] The Office
    - [ ] Aquaman
    - [ ] Pearl Jam
    - [ ] Lost

In [90]:
## Your code below


## -- Solution code

## WHERE statements

We've seen how to use do some of the most common SQL operations this far. In particular, we have:

- Learned how to write `SELECT` statements  
- Use `index_col` in the `pd.read_sql_query()` method  
- SQL Join operations
- Use SQL Aliases

In the following example, we'll look at one more technique in the SQL arsenal: the `WHERE` clause

A `WHERE` clause is followed by a **condition**. If we want to query for all invoices where country of the billing address is Germany, we can add a `Where` clause to our sql query string:

In [108]:
germany = pd.read_sql_query("SELECT * FROM invoices WHERE BillingCountry = 'Germany'", conn)
germany.head()

InvoiceId  CustomerId          InvoiceDate           BillingAddress  \
0          1           2  2009-01-01 00:00:00  Theodor-Heuss-Straße 34   
1          6          37  2009-01-19 00:00:00         Berger Straße 10   
2          7          38  2009-02-01 00:00:00      Barbarossastraße 19   
3         12           2  2009-02-11 00:00:00  Theodor-Heuss-Straße 34   
4         29          36  2009-05-05 00:00:00       Tauentzienstraße 8   

  BillingCity BillingState BillingCountry BillingPostalCode  Total  
0   Stuttgart         None        Germany             70174   1.98  
1   Frankfurt         None        Germany             60316   0.99  
2      Berlin         None        Germany             10779   1.98  
3   Stuttgart         None        Germany             70174  13.86  
4      Berlin         None        Germany             10789   1.98

`WHERE` conditions can be combined with `AND`, `OR` and `NOT`. Supposed we want to create a DataFrame containing all invoices where the billing country is **not** Indonesia, we can do the following:

In [109]:
international = pd.read_sql_query("SELECT * FROM invoices WHERE NOT BillingCountry = 'Indonesia'", conn)
international.head()

InvoiceId  CustomerId          InvoiceDate           BillingAddress  \
0          1           2  2009-01-01 00:00:00  Theodor-Heuss-Straße 34   
1          2           4  2009-01-02 00:00:00         Ullevålsveien 14   
2          3           8  2009-01-03 00:00:00          Grétrystraat 63   
3          4          14  2009-01-06 00:00:00           8210 111 ST NW   
4          5          23  2009-01-11 00:00:00          69 Salem Street   

  BillingCity BillingState BillingCountry BillingPostalCode  Total  
0   Stuttgart         None        Germany             70174   1.98  
1        Oslo         None         Norway              0171   3.96  
2    Brussels         None        Belgium              1000   5.94  
3    Edmonton           AB         Canada           T6G 2C7   8.91  
4      Boston           MA            USA              2113  13.86

**Discussion:**


Let's try by first constructing a DataFrame using the `read_sql_query()` method that we've grown familiar to. We want to develop a simple sales visualization report of our top 5 key markets (`Country` column in `customers`) ranked by Sales (`Total` column in `invoices`). 

We also want to identify our top 5 customers by name (`FirstName`, `LastName`) in the report.

Last but not least, we want the report to include a day-of-week analysis on sales performance, and for that we will need the `InvoiceDate` column. 

> **Hint 1**: `pandas` has built-in methods of extracting the name of day in a week. We've seen this in Part 2 of this specialization (**Working with Datetime chapter**). An example usage is:
>
> `x['InvoiceDOW'] = x['InvoiceDate'].dt.weekday_name`
>
>  **Hint 2**: In `read_sql_query`, you can use the `parse_dates='InvoiceDate'` argument to have the specified column parsed as date, saving you from a `to_datetime()` conversion

In [93]:
## Your code below


## -- Solution code

1. Which of the following customers are among the top 5 customers ranked by their respective total (`Total`)? Choose all that apply.
    - [ ] Victor Stevens
    - [ ] Hugh O'Reilly
    - [ ] Puja Srivastava
    - [ ] Ellie Sullivan

2. Which of the following countries are among the top 5 grossing ranked by their respective total (`Total`)? Choose all that apply.
    - [ ] United Kingdom
    - [ ] Germany
    - [ ] Portugal
    - [ ] India
    
3. How many sales transactions (or records, or rows) were from the top 5 countries?
    - Answer:

In [94]:
## Your code below


## -- Solution code

Use the code block below as reference to create a DataFrame containing records where the customers are from one of the top 5 countries. The number of rows in this new DataFrame should be equal to your answer in question (3). Notice that we have additionally converted our day-of-week column into a `Categorical` column:

```
top5 = x.groupby('Country').Total.sum().sort_values(ascending=False).head().index.to_list()
invo_topcountry = invo[invo['Country'].isin(top5)].copy()
# Represents a categorical variable in classic R fashion (pandas documentation) and have it ordered
invo_topcountry['InvoiceDOW'] = pd.Categorical(invo_topcountry.InvoiceDOW, categories=['Monday', 'Tuesday','Wednesday','Thursday','Friday','Saturday', 'Sunday'], ordered=True)
invo_topcountry.shape # same as answer for question (3)
```

4. Using the `pd.pivot_table()` method, answer the next two questions
    - 4.1 How much sales did we make in total (`Total`) from the France market on a Tuesday?
    - 4.2 How much sales did we make in total (`Total`) from all the top 5 markets combined on a Friday?
  

In [95]:
## Your code below


## -- Solution code